core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -486,7 +488,37 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
         replicationPhase.run();
         final CapturingTransport.CapturedRequest[] capturedRequests = transport.capturedRequests();
         transport.clear();
+
         assertThat(capturedRequests.length, equalTo(assignedReplicas));
+        Set<String> nodesSentTo = new HashSet<>();
+        boolean executeOnReplica =
+            action.shouldExecuteReplication(clusterService.state().getMetaData().index(shardId.getIndex()).getSettings());
+        for (CapturingTransport.CapturedRequest capturedRequest : capturedRequests) {
+            // no duplicate requests
+            assertTrue(nodesSentTo.add(capturedRequest.node.getId()));
+            // the request is hitting the correct shard
+            Request replicationRequest = (Request)capturedRequest.request;
+            assertEquals(request.shardId, replicationRequest.shardId);
+        }
+
+        // requests were sent to the correct shard copies
+        List<ShardRouting> shards =
+            clusterService.state().getRoutingTable().index(shardId.getIndex()).shard(shardId.id()).shards();
+        for (ShardRouting shard : shards) {
+            if (!shard.primary() && !executeOnReplica) {
+                continue;
+            }
+            if (shard.unassigned()) {
+                continue;
+            }
+            if (!clusterService.state().getNodes().localNodeId().equals(shard.currentNodeId())) {
```

Comment:
sorry - got confused - I thought this code skips the primary. I think it will be clear if the if here would say "ShardRouting.primary() == false "